File: tophat.py
#!/usr/bin/env python

"""Aligns reads in FASTQ files to reference genome."""

import argparse
import subprocess
import sys
import glob

__author__ = 'Alexis Blanchet-Cohen'
__date__ = '2013-01-10' 

# Parse arguments
parser = argparse.ArgumentParser(description='Aligns reads in FASTQ files to reference genome.')
parser.add_argument("-g", "--genome", help="reference genome")
parser.add_argument("-s", "--stranded", help="strandedness", default="yes")
parser.add_argument("-t", "--trimmed", help="FASTQ files trimmed", default="no")
parser.add_argument("-p", "--processors", help="number of processes")
args = parser.parse_args()

#Set the path to the genome and annotations
if(args.genome == "GRCh37"): 
    args.genomePath = "/stockage/genomes/Homo_sapiens/Ensembl/GRCh37"

elif(args.genome == "hg19"): 
    args.genomePath = "/stockage/genomes/Homo_sapiens/UCSC/hg19"

elif(args.genome == "NCBIM37"):
    args.genomePath = "/stockage/genomes/Mus_musculus/Ensembl/NCBIM37"

elif(args.genome == "GRCm38"):
    args.genomePath = "/stockage/genomes/Mus_musculus/Ensembl/GRCm38"

elif(args.genome == "mm10"):
    args.genomePath = "/stockage/genomes/Mus_musculus/UCSC/mm10"

elif(args.genome == "sacCer3"):
    args.genomePath = "/stockage/genomes/Saccharomyces_cerevisiae/UCSC/sacCer3"

elif(args.genome == "dm3"):
    args.genomePath = "/stockage/genomes/Drosophila_melanogaster/UCSC/dm3"

elif(args.genome == "BDGP5.25"):
    args.genomePath = "/stockage/genomes/Drosophila_melanogaster/Ensembl/BDGP5.25"

elif(args.genome == "ce10"):
    args.genomePath = "/stockage/genomes/Caenorhabditis_elegans/UCSC/ce10"
 
elif(args.genome == "umd3"):
    args.genomePath = "/stockage/genomes/Bos_taurus/Ensembl/UMD3.1"

else:
    sys.exit("The genome " + args.genome + " does not exist")    

os.mkdir("tophat")
os.chdir("tophat");

# Recover the filenames and the corresponding filenames.
sampleNamesFile = open ("../sampleNames.txt");

fileNames = []
sampleNames = []

for line in sampleNamesFile:
    line = line.rstrip()      # Remove trailing whitespace.
    if line:                  # Only process non-empty lines.
        fields = line.split('\t');
        fileNames.append(fields[0])
        samplesNames.append(fields[1]

sampleNamesFile.close()

##########
# TopHat #
##########

# Cycle through all the sample names.
for i in range(0,fileNames.length()/2):

    sampleName=sampleNames[i*2];

    fileNameR1 = fileNames[i*2];
    fileNameR2 = fileNames[i*2+1];

    # Make the sample directory for the TopHat output
    os.mkdirs("../../tophat/" + sampleName);

    tophatScript = open(tophatScriptName, "w")

    tophatScript.write("tophat2 --rg-library \"L\" --rg-platform \"ILLUMINA\" --rg-platform-unit \"X\" --rg-sample \"" + sampleName + "\" --rg-id \"runX\"");

    if (args.genome == "ce10"):
        tophatScript.write(" --min-intron-length 30 --max-intron-length 30000");

    tophatScript.write(" --no-novel-juncs -p " + args.processors); 

    if (stranded == "stranded"):
        tophatScript.write(" --library-type fr-firststrand");

    tophatScript.write(" -G args.genomePath/Annotation/Genes/genes.gtf");  
    tophatScript.write(" -o ../../tophat/" + sampleName);
    tophatScript.write(" args.genomePath/Sequence/Bowtie2Index/genome");

    if (args.trimmed == "untrimmed"):
        tophatScript.write(" ../../../FASTQ/untrimmed/" + fileNameR1);
        tophatScript.write(" ../../../FASTQ/untrimmed/" + fileNameR2);
        
    else:
        tophatScript.write(" ../../../FASTQ/trimmed/" + fileNameR1);
        tophatScript.write(" ../../../FASTQ/trimmed/" + fileNameR2);  

    tophatScript.write(TOPHAT " &> " + tophatScriptName + ".sh_output");

    tophatScript.close()

subprocess.call("submitJobs.py");

File: tophat.pl
#!/usr/bin/perl

#####################################################
# Generates the Tophat scripts.                     #
# Usage: tophat.pl genomeName stranded|unstranded   # 
# e.g. tophat.pl hg19 unstranded                    #
#                                                   # 
# Author: Alexis Blanchet-Cohen                     #
#####################################################

use strict;
use warnings;

my $genome;
my $genomePath;

my $numberProcessors;
my $stranded;
my $trimmed;

my @sampleNames;
my $sampleName;

my @fileNames;
my $fileNameR1;
my $fileNameR2;

my $tophatScriptName;

my $hostname;
my $host;

my $genomesFolder = "/stockage/genomes";

# Check the number of arguments.
if ($#ARGV+1 != 4) {
  print "Wrong number of arguments.\n";
  print "Usage: tophat.pl hg19 stranded|unstranded trimmed|untrimmed numberProcessors\n";
  print "e.g. tophat.pl hg19 stranded trimmed 12\n";
  exit;
}

# Read the command line argument.
$genome = $ARGV[0];
$stranded = $ARGV[1];
$trimmed = $ARGV[2];
$numberProcessors = $ARGV[3];

#Set the path to the genome and annotations
if($genome eq "GRCh37") { 
  $genomePath = "$genomesFolder/Homo_sapiens/Ensembl/GRCh37";
}
elsif($genome eq "hg19") { 
  $genomePath = "$genomesFolder/Homo_sapiens/UCSC/hg19";
}
elsif ($genome eq "NCBIM37") {
  $genomePath = "$genomesFolder/Mus_musculus/Ensembl/NCBIM37";
}
elsif ($genome eq "GRCm38") {
  $genomePath = "$genomesFolder/Mus_musculus/Ensembl/GRCm38";
}
elsif ($genome eq "mm10") {
  $genomePath = "$genomesFolder/Mus_musculus/UCSC/mm10";
}
elsif ($genome eq "sacCer3") {
  $genomePath = "$genomesFolder/Saccharomyces_cerevisiae/UCSC/sacCer3";
}
elsif ($genome eq "dm3") {
  $genomePath = "$genomesFolder/Drosophila_melanogaster/UCSC/dm3";
} 
elsif ($genome eq "BDGP5.25") {
  $genomePath = "$genomesFolder/Drosophila_melanogaster/Ensembl/BDGP5.25";
} 
elsif ($genome eq "ce10") {
  $genomePath = "$genomesFolder/Caenorhabditis_elegans/UCSC/ce10";
} 
elsif ($genome eq "umd3") {
  $genomePath = "$genomesFolder/Bos_taurus/Ensembl/UMD3.1";
} 
else {
  print "genome: " . $genome;
  print "Usage: tophat.pl hg19 stranded|unstranded trimmed|untrimmed numberProcessors\n";
  print "e.g. tophat.pl hg19 stranded trimmed 12\n";
  exit;  
}

`mkdir tophat`;
chdir("tophat");

# Recover the filenames and the corresponding filenames.
open (MYFILE, '../sampleNames.txt');

while (<MYFILE>) {
	chomp($_);
	next if /^\s*$/; # Skip blank lines. 
	my @fields = split('\t', $_);
	push(@fileNames,$fields[0]);
	push(@sampleNames,$fields[1]);	 
}

close (MYFILE); 

##########
# TopHat #
##########

# Cycle through all the sample names.
for (my $i = 0; $i < scalar(@fileNames)/2; $i++) {

  $sampleName=$sampleNames[$i*2];

  $fileNameR1 = $fileNames[$i*2];
  $fileNameR2 = $fileNames[$i*2+1];

  # Make the sample directory for the TopHat output
  `mkdir -p ../../tophat/$sampleName`;

  $tophatScriptName = "tophat_" . $sampleName;
  my $tophatScriptName_handle = ">" . $tophatScriptName . ".sh";

  open (TOPHAT, $tophatScriptName_handle) or die "Unable to open $tophatScriptName";

  print TOPHAT "tophat2 --rg-library \"L\" --rg-platform \"ILLUMINA\" --rg-platform-unit \"X\" --rg-sample \"$sampleName\" --rg-id \"runX\"";
  if ($genome eq "ce10") {
    print TOPHAT " --min-intron-length 30 --max-intron-length 30000";
  }
  print TOPHAT " --no-novel-juncs -p $numberProcessors"; 
  if ($stranded eq "stranded") {
    print TOPHAT " --library-type fr-firststrand";
  }
  print TOPHAT " -G $genomePath/Annotation/Genes/genes.gtf";  
  print TOPHAT " -o ../../tophat/$sampleName";
  print TOPHAT " $genomePath/Sequence/Bowtie2Index/genome";
  if ($trimmed eq "untrimmed") {
    print TOPHAT " ../../../FASTQ/untrimmed/$fileNameR1";
    print TOPHAT " ../../../FASTQ/untrimmed/$fileNameR2";
  } 
  else {
    print TOPHAT " ../../../FASTQ/trimmed/$fileNameR1";
    print TOPHAT " ../../../FASTQ/trimmed/$fileNameR2";  
  } 

  print TOPHAT " &> $tophatScriptName\.sh_output";


  close(TOPHAT)
}

`submitJobs.py`;
